hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java

```diff
@@ -348,14 +436,85 @@ private void processSnapshotDeepClean(int delCount)
               if (previousSnapshot != null) {
                 rcPrevOmSnapshot.close();
               }
+              if (previousToPrevSnapshot != null) {
+                rcPrevToPrevOmSnapshot.close();
+              }
             }
           }
 
         }
       }
+
+      updateSnapshotExclusiveSize();
       updateDeepCleanedSnapshots(deepCleanedSnapshots);
     }
 
+    private OmKeyInfo getPreviousSnapshotKeyName(
+        OmKeyInfo keyInfo, OmBucketInfo bucketInfo, long volumeId,
+        Table<String, String> snapRenamedTable,
+        Table<String, OmKeyInfo> previousKeyTable) throws IOException {
+
+      if (keyInfo == null) {
+        return null;
+      }
+
+      String dbKeyPrevSnap;
+      if (bucketInfo.getBucketLayout().isFileSystemOptimized()) {
+        dbKeyPrevSnap = getOzoneManager().getMetadataManager().getOzonePathKey(
+            volumeId,
+            bucketInfo.getObjectID(),
+            keyInfo.getParentObjectID(),
+            keyInfo.getFileName());
+      } else {
+        dbKeyPrevSnap = getOzoneManager().getMetadataManager().getOzoneKey(
+            keyInfo.getVolumeName(),
+            keyInfo.getBucketName(),
+            keyInfo.getKeyName());
+      }
+
+      String dbRenameKey = getOzoneManager().getMetadataManager().getRenameKey(
+          keyInfo.getVolumeName(),
+          keyInfo.getBucketName(),
+          keyInfo.getObjectID());
+
+      String renamedKey = snapRenamedTable.getIfExist(dbRenameKey);
+      dbKeyPrevSnap = renamedKey != null ? renamedKey : dbKeyPrevSnap;
+
+      return previousKeyTable.get(dbKeyPrevSnap);
+    }
+
+    private void updateSnapshotExclusiveSize() {
+
+      List<SnapshotProperty> snapshotPropertyList = new ArrayList<>();
+      for (String dbKey: completedExclusiveSizeList) {
+        SnapshotProperty snapshotProperty = SnapshotProperty.newBuilder()
+                .setSnapshotKey(dbKey)
+                .setExclusiveSize(exclusiveSizeList.get(dbKey))
+                .setExclusiveReplicatedSize(
+                    exclusiveReplicatedSizeList.get(dbKey))
+                .build();
+        snapshotPropertyList.add(snapshotProperty);
+        exclusiveSizeList.remove(dbKey);
+        exclusiveReplicatedSizeList.remove(dbKey);
+      }
+
+      if (!snapshotPropertyList.isEmpty()) {
```

Comment:
nit: this could be a first check in `updateSnapshotExclusiveSize()`. `completedExclusiveSizeList` is empty early return otherwise update the snapshot info.

Reply:
This check is not needed any more.

Could be simply this:
```
    private void updateSnapshotExclusiveSize() {
      if (completedExclusiveSizeMap.isEmpty()) {
        return;
      }

      List<SnapshotProperty> snapshotPropertyList = new ArrayList<>();

      for (String dbKey: completedExclusiveSizeMap) {
        SnapshotProperty snapshotProperty = SnapshotProperty.newBuilder()
                .setSnapshotKey(dbKey)
                .setExclusiveSize(exclusiveSizeMap.get(dbKey))
                .setExclusiveReplicatedSize(
                    exclusiveReplicatedSizeMap.get(dbKey))
                .build();
        snapshotPropertyList.add(snapshotProperty);
        exclusiveSizeMap.remove(dbKey);
        exclusiveReplicatedSizeMap.remove(dbKey);
      }

      SetSnapshotPropertyRequest setSnapshotPropertyRequest =
          SetSnapshotPropertyRequest.newBuilder()
              .addAllSnapshotProperty(snapshotPropertyList)
              .build();

      OMRequest omRequest = OMRequest.newBuilder()
          .setCmdType(Type.SetSnapshotProperty)
          .setSetSnapshotPropertyRequest(setSnapshotPropertyRequest)
          .setClientId(clientId.toString())
          .build();

      submitRequest(omRequest);
      completedExclusiveSizeMap.clear();
    }
```